src/v/config/property_constraints.cc

```diff
@@ -0,0 +1,462 @@
+                       : model::compression::none;
+          },
+          [](
+            cluster::topic_configuration& topic_cfg,
+            model::compression& topic_val) {
+              topic_cfg.properties.compression = topic_val;
+          });
+    } else if (name == "kafka_batch_max_bytes") {
+        std::optional<uint32_t> min_opt{std::nullopt}, max_opt{std::nullopt};
+        if (min_str_opt) {
+            min_opt = static_cast<uint32_t>(std::stoul(*min_str_opt));
+        }
+        if (max_str_opt) {
+            max_opt = static_cast<uint32_t>(std::stoul(*max_str_opt));
+        }
+        auto def_val = cfg.kafka_batch_max_bytes.default_value();
+        return std::make_shared<constraint_validator_range<uint32_t>>(
+          std::move(min_opt),
+          std::move(max_opt),
+          [def_val{std::move(def_val)}](
+            const cluster::topic_configuration& topic_cfg) {
+              return topic_cfg.properties.batch_max_bytes
+                       ? *topic_cfg.properties.batch_max_bytes
+                       : def_val;
+          },
+          [](cluster::topic_configuration& topic_cfg, uint32_t& topic_val) {
+              topic_cfg.properties.batch_max_bytes = topic_val;
+          });
+    } else if (name == "log_message_timestamp_type") {
+        auto def_val = cfg.log_message_timestamp_type.default_value();
+        return std::make_shared<
+          constraint_validator_enabled<model::timestamp_type>>(
+          std::move(enabled),
+          cfg.log_message_timestamp_type.bind(),
+          [def_val{std::move(def_val)}](
+            const cluster::topic_configuration& topic_cfg) {
+              return topic_cfg.properties.timestamp_type
+                       ? *topic_cfg.properties.timestamp_type
+                       : def_val;
+          },
+          [](
+            cluster::topic_configuration& topic_cfg,
+            model::timestamp_type& topic_val) {
+              topic_cfg.properties.timestamp_type = topic_val;
+          });
+    } else if (name == "cloud_storage_enable_remote_read") {
+        return std::make_shared<constraint_validator_enabled<bool>>(
+          std::move(enabled),
+          cfg.cloud_storage_enable_remote_read.bind(),
+          [](const cluster::topic_configuration& topic_cfg) {
+              return topic_cfg.properties.shadow_indexing
+                       ? model::is_fetch_enabled(
+                         *topic_cfg.properties.shadow_indexing)
+                       : false;
+          },
+          [](cluster::topic_configuration& topic_cfg, bool& topic_val) {
+              // The broker will assign fetch for remote read in AlterConfigs
+              // and IncrementalAlterConfigs handlers. Therefore, we do the same
+              // here.
+              topic_cfg.properties.shadow_indexing
+                = topic_val ? model::shadow_indexing_mode::fetch
+                            : model::shadow_indexing_mode::disabled;
```

Comment:
question: if enable_remote_read and enable_remote_write are enabled, the value of this should be `model::shadow_indexing_mode::full` right?  Won't this overwrite that?

Reply:
Asking the storage team about this in https://redpandadata.slack.com/archives/C04ASNM2ZLK/p1697136266997769